Store the backup JPEG images in `_wp_attachment_backup_sizes`

mitogh opened the pull request:
## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->

Creates the JPEG version of the WebP images and store the references into `_wp_attachment_backup_sizes` similar to image transformations in order to take advantage of benefits from WP Core like files removed automatically by WordPress and uploaded into CDNs

Fixes:

- #142 
- #150 

## Relevant technical choices

The changes introduced in this PR replicate the same behavior from WordPress core to store edited versions of additional mime types and the additional mime types themselves in the meta `_wp_attachment_backup_sizes`.



## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

mitogh committed:
Add the `sources` property to each image size

The property is added to every single image size available,
each value inside the properties array is a reference to
a different image mime type.

If the image size was created as a JPEG version a WebP version
would be created, if a WebP version exists a JPEG version
would be created instead based on the details from ticket #142

This mechanism would allow extending the sources properties for
more additional mime types.

The value for each mime type is an associative array with `file` as
the only property for now. And the plan is to include more meaningful
properties such as `filesize` to a different mime type.

mitogh committed:
Fix linting errors from `phpcs`

Make sure file follows the standard from `phpcs`

mitogh committed:
Make sure the correct size is updated

Prevent to use numeric indexes for each image size, instead use the
correct image name.

mitogh committed:
Make sure the `scaled` version is used.

In this case the `_wp_attached_file` needs to be used in order
to gain any update to the image, instead of using the original
image for any subsize.

mitogh committed:
Generate all the backup images

Create an store the backup images for all the subsizes
and supported image sizes.

mitogh committed:
Fix `phocs` standard guidelines

Make sure the file is formatted properly.

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -35,5 +35,170 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
+			// Remove the path of the image to follow the same pattern as core.
+			unset( $image['path'] );
+
+			$backup_sizes[ $size . '_' . $extension ] = $image;
```
```suggestion
			$backup_sizes[ $size . '-' . $extension ] = $image;
```
we use a dash everywhere else for filename modification

mitogh replied:
Thanks, great flag.

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -35,5 +35,170 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
+			}
+
+			// TODO: Store the file size of the created image.
+			// $image['filesize'] = filesize( $image['path'] );
```
I'm in favor of storing this for all sizes, this is useful for performance analysis and potentially other things.

mitogh replied:
Great. for now would keep the TODO, and we can came back to this once we start working on: 

- https://github.com/WordPress/performance/issues/8

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -35,5 +35,170 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
+ * @return string[] An array of valid mime types, where the key is the mime type and the value is the extension type.
+ */
+function webp_uploads_valid_image_mime_types() {
+	return array(
```
maybe should be filtered here?

mitogh replied:
Agree, added a todo at the top of the function, but let me add the filter right away instead.

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -35,5 +35,170 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
+	$backup_sizes = get_post_meta( $attachment_id, '_wp_attachment_backup_sizes', true );
+	$backup_sizes = is_array( $backup_sizes ) ? $backup_sizes : array();
+
+	foreach ( webp_uploads_get_image_sizes() as $size => $properties ) {
```
instead of trying to recreate sizes, can we just copy the meta 'sizes'?

mitogh replied:
We can certainly do that but aside from that, we need to know if the newly created image should be `cropped` or not and that's the main reason why we need this function. 

I think something we can do at the top of the `foreach` loop to make sure this `$size` exists in the original metadata before creating a targeted image instead.

mitogh replied:
Actually, this is already the case on this line :smiley:  

https://github.com/WordPress/performance/blob/84eb3ec30e532939cf81c65d4fc46120ae1aac8c/modules/images/webp-uploads/load.php#L81

adamsilverstein replied:
One thing I noticed in testing when I uploading a very large image, I got a jpeg created with a `-scaled` suffix, then **all** of the generated `.webp` sub size files also included the `-scaled` suffix:

![image](https://user-images.githubusercontent.com/2676022/154143125-397cc077-76ca-40e3-a57f-abef530a206c.png)

In addition, I noticed there was no full sized `.webp` created (that is `{filename}-scaled.webp`) - we need that full sized file for the image markup (only applies when you upload images larger than `big_image_size_threshold`)

mitogh replied:
Thanks for testing and for the feedback. Some follow-up comments.

> One thing I noticed in testing when I uploading a very large image, I got a jpeg created with a -scaled suffix, then all of the generated .webp sub size files also included the -scaled suffix.

Great point mostly because this is created out of the scaled version. But let me fix this one :) 

> In addition, I noticed there was no full sized .webp created (that is {filename}-scaled.webp) - we need that full sized file for the image markup (only applies when you upload images larger than big_image_size_threshold)

Actually, this follows the same logic applied here #122  and #143   this behavior would go against this solved issue, as this `scaled` version is the `full` size image. Based on the issue `scaled` and `full` size always remains in the original image format. (not converted)

Let me know if you think we should change it instead.

mitogh committed:
Use dash for file name modification.

Co-authored-by: Adam Silverstein <adamjs@google.com>

mitogh committed:
Create additional mime types when an image is edited.

Update the logic for hwo the additional mime types were created
in order to allow image edits to take effect, when an image is
cropped, flipped or rotated, those transformations are reflected
into the additional mime types.

eugene-manuilov commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,36 +4,434 @@
+ *
+ * @return string A function that would be used as a callback.
+ */
+function webp_uploads_wp_die_ajax_handler() {
+	return 'webp_uploads_ajax_wp_die_handler';
+}
```
This hook should execute the callback function passed to it via filter instead of calling the `_ajax_wp_die_handler` function:

```php
function webp_uploads_wp_die_ajax_handler( $callback ) {
	return function( $message, $title = '', $args = array() ) use ( $callback ) {
		if ( ... ) {
			webp_uploads_restore_image_from_backup( absint( $_REQUEST['postid'] ) );
		}

		call_user_func( $callback, $message, $title, $args );
	};
}
```

Also when you register this hook it should use the biggest priority number to make sure it is executed at the very end:

```php
add_filter( 'wp_die_ajax_handler', 'webp_uploads_wp_die_ajax_handler', PHP_INT_MAX );
```

mitogh replied:
> This hook should execute the callback function passed to it via filter instead of calling the _ajax_wp_die_handler function:

Right, great point let me update the code accordingly :) 

> Also when you register this hook it should use the biggest priority number to make sure it is executed at the very end:

Well in this case we know even if is executed first on the set of hooks won't make any difference so the default seems reasonable. 

Ideally we should have a filter in the ajax via core for restore and edited images.

mitogh committed:
Update key structure to be `extension-type`

This would allow to quickly locate `webp` or a defined extension
images and split between `-orig` images suffix as well.

mitogh committed:
Rename function to avoid namespace collisions.

Prevent namespace collisions when using the function
from within the plugin.

mitogh committed:
Return the provided `callback` if present

The callback was passed as a parameter, no change
is placed on the callback so we can safely return it.

mitogh committed:
Update logic to perform updates on the additional mimes

This was due in some cases when restoring an image and editing
it again, the image would result in not executing correctly
due the meta was the same as before and the update callback was
not executed, instead we are moving this work into an ajax callback.

mitogh committed:
Merge branch 'trunk' into feature/142-multiple-mimes-for-image-sizes-with-backup

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
+	}
+
+	// All subsizes are created out of the `file` property and not the original image.
+	$file = get_attached_file( $attachment_id, true );
```
Why do you use get_attached_file here instead of the original image, is this to use the -scaled image?

mitogh replied:
Correct any edits to an image are performed to the `scaled` version (if present) by WordPress so any subsize should be based out of the modified image and not the original image instead. 

The original image is only used to create the scaled version if not required it would be used as the attached file instead.

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
+			// The file already exists as part of the backup sizes in the same key.
+			if ( array_key_exists( $key, $backup_sizes ) ) {
+				// This case would remove the file if is an edited version.
+				if ( defined( 'IMAGE_EDIT_OVERWRITE' ) && IMAGE_EDIT_OVERWRITE ) {
```
Nice!

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
+					$file_name     = empty( $backup_sizes[ $key ]['file'] ) ? '' : $backup_sizes[ $key ]['file'];
+					$file_location = path_join( $dirname, $file_name );
+					// Test to make sure this is an edited version, if so we can remove it safely.
+					preg_match( '/-e[\d]{13}-/', $file_name, $matches );
```
consider storing this regex in a constant since you reuse it

mitogh replied:
Great point, let me do that instead.

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
+				unset( $backup_sizes[ $key ] );
+			}
+
+			// Editor needs to be recreated every time as there is not flush() or clear() function that can be used after we created an image.
```
Did you try`$editor->__destruct` (or even just $editor = null which should trigger `__destruct`)?  For both Imagick and GD this should clear the image.

mitogh replied:
No, let me try that thanks for the suggestion.

mitogh replied:
Actually calling `__destruct` seems like the same as creating a new variable altogether as essentially this is called by the engine itself so, for now, will keep it like that. 

One thing that I've noted out of the class `WP_Image_Editor` is the fact that has a property `get_size` but not a `set_size` which would be the one we can use here to reset the size to the appropriate dimensions. Any thoughts on that one ?

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
+	// This should take place only on the JPEG image.
+	$valid_mime_types = webp_uploads_valid_image_mime_types();
+
+	// Not a supported mime type to create the sources' property.
```
stray `'` in this comment

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
- * @since 1.0.0
  */
 
+add_filter( 'wp_generate_attachment_metadata', 'webp_uploads_create_images_with_additional_mime_types', 10, 3 );
```
Generating the additional mime types immediately might fail because it will double execution time, right? I can test this locally, but I think this might break uploads for some users.

Can we instead add an immediately one-time executing cron callback (`wp_schedule_single_event`)? that way the images will get generated on the next callback as part of a separate request?

mitogh replied:
Agree I'm going to explore this as part of enchantment on top of this PR [as you describe](https://github.com/WordPress/performance/pull/154#issuecomment-1040833914) has become quite large at this point.

mitogh replied:
Actually I just updated this PR to handle this case.

adamsilverstein commented on modules/images/webp-uploads/load.php:
```diff
@@ -4,41 +4,414 @@
+	);
+
+	/**
+	 * An array representing all the valid mime types where multiple images would be created if
```
We should make sure that by returning only `webp` in the array **only webp images are generated** - that is, the uploads would work the way they do now in the plugin. 

Although we decided this shouldn't be the default approach, it may turn out to be the best approach for most users, and over time it may become our choice for default. Code wise, we should be able to keep the existing plugin approach, but only use it in the case when a single mime type is specified.

mitogh replied:
Great point. 

Let me think about how I can go about this use case when only a single mime type is present either JPEG or WebP so I can update this PR accordingly. Another use case we need to consider is the fact that due this is filterable it can be empty as well. Under that scenario we can either:

- Generate all subsizes with the same mime-type as the original image regardless if this filter is empty
- Not generate any subsize to follow what was specified on the filter.

It feels like the last point would be the way to go for consistency but let me know your thoughts on that one.

mitogh replied:
Fixed this one and added a TODO to handle this last point.

adamsilverstein commented:
@mitogh this looks good so far, I left some small comments and questions.

This PR has grown quite large because in addition to handling image generation you are supporting image edit and restore. I realize these are related because of the data storage, but still wonder if we could break them out so we can focus solely on the image generation and meta data storage. Once we land that, we can tackle the edit/restore bits in a separate PR.

mitogh committed:
Remove non required character.

mitogh committed:
Move hash parsing into a custom function.

THe hash extraction ahs been moved into a custom function to avoid
confusion and bring clarity when used.

mitogh committed:
Remove suffix from file names.

Remove the suffix:

- scaled
- rotated
- imagifyresized

From the file names if present.

mitogh committed:
Split the edit logic into a separate Pull Request

The logic to handle the case for image edits like: rotation, flip
and crop would be handled in a separate PR.

This PR addresses the concern for #142 and #150.

mitogh committed:
Format function declaration and parameters

mitogh committed:
Simplify mock classes with an inline return instead.

The result of the condition is always a boolean, so this
section can be simplified with a single line instead.

mitogh committed:
Move the generation of additional images into a cron.

Additional images would be created during a cron execution,
removing any workload from the initial request.

Tests where fixed and some new ones where introduced.

mitogh committed:
Fix linter errors on test file

felixarntz commented:
@mitogh @adamsilverstein I think the approach in #147 makes more sense than this one. `_wp_attachment_backup_sizes` is used for edited/cropped versions of an image, while the `sizes` field of the regular image metadata is what is being used for the registered sizes. Since we are focusing on generating the sub-sizes in progressive image formats, I think our data should be in the `sizes` array.

Regarding backward-compatibility, I don't think the approach here justifies going with it given the unexpected data storage, specifically for the following reasons:
* It shouldn't be too hard deleting the "custom" sub-sizes generated in the approach from #147, we just have to use the `delete_attachment` action to delete any of those image files that exist.
* Regarding CDNs, that indeed is a valid concern. However, we also don't break anything in these plugins or WordPress core with the `sources` implementation in #147, and the eventual dev note post about this feature in WordPress core should include guidance on how to update such type of plugins to support the new `sources` field. It's okay to introduce a new API for plugins to use, and since we're adding a new feature to WordPress core, it's reasonable to expect plugin developers that want to support the feature to adopt the new "API" (which here really is just newly stored data).

mitogh commented:
Closing this one in favor of: 

- #147 

Details https://github.com/WordPress/performance/pull/147#issuecomment-1043145067
